### PR TITLE
Bugfix/259 smaller issues around create delta pkg and deploy

### PR DESCRIPTION
**Pull request by JoernBerkefeld**
# PR details

## What is the purpose of this pull request? (put an "X" next to an item)

_Please delete options that are not relevant._

- [ ] Documentation update
- [x] Bug fix
- [ ] New metadata support
- [ ] Enhanced metadata
- [ ] Add a CLI option
- [ ] Add something to the core
- [ ] Other, please explain:

## What changes did you make? (Give an overview)

...

## Is there anything you'd like reviewers to focus on?

...

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] ESLint & Prettier are not outputting errors or warnings
- [ ] README.md updated (if applicable)
- [ ] CHANGELOG.md updated
- [x] ran `npm run docs` to update developer docs

### CODE REVIEW

File: lib/util/util.js [hunk 1]
```diff
@@ -50,6 +50,31 @@ const Util = {
...
+    /**
+     * helper to quickly ensure we are working with a flat Array of items
+     * @param {MetadataTypeMap|MetadataTypeItem[]} metadata map or array of metadata
+     * @returns {MetadataTypeItem[]} Array of metadata items
+     */
+    getMetadataMapToList(metadata) {
+        if (metadata && Array.isArray(metadata)) {
+            return metadata;
+        } else {
+            return Object.keys(metadata).map((item) => metadata[item]);
```
Comment (DougMidgley) [inline lib/util/util.js:62]: I think you mean to use: 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/values
Also, I would wonder if this is really the right approach? We should always know what we are returning, and if this is needed it shows there is bugs which need fixing

File: lib/index.js [hunk 2]
```diff
@@ -121,6 +121,7 @@ class Mcdev {
                 // Other map only with type -> file (for buildDefinitionBulk)
                 const buTypeDelta = {};
                 const typeDelta = {};
+                let deltaCounter = 0;
```
Comment (DougMidgley) [inline lib/index.js:124]: not loving increasing logic creep in what is the index.js

File: lib/metadataTypes/Asset.js [hunk 3]
```diff
@@ -101,7 +101,7 @@ class Asset extends MetadataType {
         }
         Util.logger.info(`Downloaded: ${this.definition.type} (${Object.keys(metadata).length})`);
 
-        return { metadata: metadata, type: this.definition.type };
+        return { metadata: metadata[Object.keys(metadata)[0]], type: this.definition.type };
```
Comment (DougMidgley) [inline lib/metadataTypes/Asset.js:104]: Also wondering if we shouldnt be doing this earlier? 
```suggestion
        return { metadata: Object.values(metadata)[0], type: this.definition.type };
```

File: lib/util/util.js [hunk 4]
```diff
@@ -50,6 +50,31 @@ const Util = {
...
+        }
+    },
+    /**
+     * helper to turn an array or object into a single item
+     * @param {Object.<{metadata:MetadataTypeMap,type:string}>} obj map or array of metadata
+     * @returns {Object.<{metadata:MetadataTypeItem,type:string}>} item list reduced to single item
+     */
+    reduceMetadataMap(obj) {
```
Comment (DougMidgley) [inline lib/util/util.js:70]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/values
also, i wonder if the name like "extractSingleObject" wouldnt be better? reduce has a specific meaning in js, so may be better to avoid, since it isnt reducing like you might expect

File: lib/metadataTypes/Script.js [hunk 5]
```diff
@@ -53,15 +53,17 @@ class Script extends MetadataType {
...
      */
     static async retrieveAsTemplate(templateDir, name, templateVariables) {
         await File.initPrettier('ssjs');
-        return super.retrieveREST(
-            templateDir,
-            '/automation/v1/scripts/?$filter=name%20eq%20' + encodeURIComponent(name),
-            null,
-            templateVariables
+        return Util.reduceMetadataMap(
+            await super.retrieveREST(
```
Comment (DougMidgley) [inline lib/metadataTypes/Script.js:61]: open point for discussion if retrieveREST wouldnt be the better place to put this? we could conditionally change result based on template.

**Review by DougMidgley — CHANGES_REQUESTED** [review]

File: lib/index.js (hunk 2 above)
Comment (JoernBerkefeld) [reply lib/index.js:124]: Open to ideas on how to solve this more efficiently ;-)

File: lib/metadataTypes/Script.js (hunk 5 above)
Comment (JoernBerkefeld) [reply lib/metadataTypes/Script.js:61]: Good idea. Let me check the feasibility when I get back to my laptop (travelling atm)

File: lib/util/util.js (hunk 4 above)
Comment (JoernBerkefeld) [reply lib/util/util.js:70]: renamed to extractSingleItem()

Comment (JoernBerkefeld) [reply lib/util/util.js]: removed method

File: lib/metadataTypes/Script.js (hunk 5 above)
Comment (JoernBerkefeld) [reply lib/metadataTypes/Script.js]: done

**Comment by JoernBerkefeld** [comment]
final review please @DougMidgley 
should be all good now

File: lib/metadataTypes/Discovery.js [hunk 6]
```diff
@@ -33,7 +33,7 @@ class Discovery extends MetadataType {
                 v.key = Object.keys(this.definition.endPointMapping)[i];
                 metadataStructure[v.key] = v;
             }
-            await this.saveResults(metadataStructure, retrieveDir, null);
+            await super.saveResults(metadataStructure, retrieveDir, null);
```
Comment (DougMidgley) [inline lib/metadataTypes/Discovery.js:36]: why did you move to super?

Comment (JoernBerkefeld) [reply lib/metadataTypes/Discovery.js:36]: as discussed, just to signal more clearly that this is actually using the super-version and not a local one. like we've done it in other type-classes.
